include/output.hxx

```diff
@@ -50,64 +49,211 @@ using std::endl;
+    if (enabled) {
+      base->vwrite(str, va);
+    }
+  }
+
+  /// If enabled, print a string to stdout using C printf formatting
+  /// note: unlike write, this is not also sent to log files
+  void print(const char *str, ...) override;
+  void vprint(const char *str, va_list va) override {
+    if (enabled) {
+      base->vprint(str, va);
+    }
+  }
+  
+  /// Get the Output object which is the base of this ConditionalOutput
+  Output *getBase() {
+    if (base_is_cond) {
+      return dynamic_cast<ConditionalOutput *>(base)->getBase();
+    } else {
+      return base;
+    }
+  };
   
-  std::ofstream file; ///< Log file stream
-  static const int BUFFER_LEN=1024; ///< default length
-  int buffer_len; ///< the current length
-  char * buffer; ///< Buffer used for C style output
-  bool enabled;      ///< Whether output to stdout is enabled
+  /// Set whether this ConditionalOutput is enabled
+  /// If set to false (disabled), then all print and write calls do nothing
+  void enable(bool enable_) { enabled = enable_; };
+
+  /// Turn on outputs through calls to print and write
+  void enable() override { enabled = true; };
+
+  /// Turn off outputs through calls to print and write
+  /// This includes log files and stdout
+  void disable() override { enabled = false; };
+
+  /// Check if output is enabled
+  bool isEnabled() {
+    return enabled &&
+           (!base_is_cond || (dynamic_cast<ConditionalOutput *>(base))->isEnabled());
+  };
+
+  Output *base;
+  bool enabled;
+
+private:
+  bool base_is_cond;
+
+private:
+};
+
+/// Catch stream outputs to DummyOutput objects. This is so that
+/// statements like
+///    output_debug << "debug message";
+/// compile but have no effect if DEBUG_ENABLED is false
+template <typename T> DummyOutput &operator<<(DummyOutput &out, T const &t) {
+  return out;
+}
+
+template <typename T> DummyOutput &operator<<(DummyOutput &out, const T *t) {
+  return out;
+}
+
+inline DummyOutput &operator<<(DummyOutput &out, std::ostream &(*pf)(std::ostream &)) {
+  return out;
+}
+
+inline ConditionalOutput &operator<<(ConditionalOutput &out,
+                                     std::ostream &(*pf)(std::ostream &)) {
+  if (out.isEnabled()) {
+    *out.getBase() << pf;
+  }
+  return out;
+};
+
+template <typename T> ConditionalOutput &operator<<(ConditionalOutput &out, T const &t) {
+  if (out.isEnabled()) {
+    *out.getBase() << t;
+  }
+  return out;
+};
+
+template <typename T> ConditionalOutput &operator<<(ConditionalOutput &out, const T *t) {
+  if (out.isEnabled()) {
+    *out.getBase() << t;
+  }
+  return out;
 };
 
 /// To allow statements like "output.write(...)" or "output << ..."
-#define output (*Output::getInstance())
+/// Output for debugging
+#ifdef DEBUG_ENABLED
+extern Output output_debug;
+#else
+extern DummyOutput output_debug;
+#endif
+extern ConditionalOutput output_warn;  ///< warnings
+extern ConditionalOutput output_prog;  ///< progress
```

Comment:
I think this should just be `output_progress`, `prog` is very ambiguous for the sake of saving 4 characters.

Reply:
All output_prog changed to output_progress